pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/gofrs/uuid"
+	"github.com/gorilla/mux"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/index"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/grafana/dskit/runutil"
+	"github.com/grafana/dskit/tenant"
+	"github.com/grafana/regexp"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	"github.com/grafana/mimir/pkg/storage/sharding"
+	mimir_tsdb "github.com/grafana/mimir/pkg/storage/tsdb"
+)
+
+// CreateBlockUpload handles requests for creating block upload sessions.
+func (c *MultitenantCompactor) CreateBlockUpload(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
```

Comment:
Applies to all endpoints: we need to validate if blockID is correct (ulid.Parse it and fail on error)

Reply:
Done.